Sparkle/SUPlainInstaller.m

```diff
@@ -7,13 +7,165 @@
+    NSURL *tempNewDirectoryURL = [fileManager makeTemporaryDirectoryWithPreferredName:[installationURL.lastPathComponent.stringByDeletingPathExtension stringByAppendingString:@" (Incomplete Update)"] appropriateForDirectoryURL:installationURL.URLByDeletingLastPathComponent error:error];
+    if (tempNewDirectoryURL == nil) {
+        SULog(@"Failed to make new temp directory");
+        return NO;
+    }
+    
+    // Move the new app to our temporary directory
+    NSURL *newTempURL = [tempNewDirectoryURL URLByAppendingPathComponent:newURL.lastPathComponent];
+    if (![fileManager moveItemAtURL:newURL toURL:newTempURL error:error]) {
+        SULog(@"Failed to move the new app from %@ to its temp directory at %@", newURL.path, newTempURL.path);
+        [fileManager removeItemAtURL:tempNewDirectoryURL error:NULL];
+        return NO;
+    }
+    
+    // Release our new app from quarantine and fix its owner and group IDs while it's at our temporary destination
+    // We must leave moving the app to its destination as the final step in installing it, so that
+    // it's not possible our new app can be left in an incomplete state at the final destination
+    
+    NSError *quarantineError = nil;
+    if (![fileManager releaseItemFromQuarantineAtRootURL:newTempURL error:&quarantineError]) {
+        // Not big enough of a deal to fail the entire installation
+        SULog(@"Failed to release quarantine at %@ with error %@", newTempURL.path, quarantineError);
+    }
+    
+    NSURL *oldURL = [NSURL fileURLWithPath:host.bundlePath];
+    if (![fileManager changeOwnerAndGroupOfItemAtRootURL:newTempURL toMatchURL:oldURL error:error]) {
+        // But this is big enough of a deal to fail
+        SULog(@"Failed to change owner and group of new app at %@ to match old app at %@", newTempURL.path, oldURL.path);
+        [fileManager removeItemAtURL:tempNewDirectoryURL error:NULL];
+        return NO;
+    }
+    
+    // Decide on a destination name we should use for the older app when we move it around the file system
+    NSString *oldDestinationName = nil;
+    if (SPARKLE_APPEND_VERSION_NUMBER) {
+        NSString *oldBundleVersion = [self bundleVersionAppropriateForFilenameFromHost:host];
+        
+        oldDestinationName = [oldURL.lastPathComponent.stringByDeletingPathExtension stringByAppendingFormat:@" (%@)", oldBundleVersion != nil ? oldBundleVersion : @"old"];
+    } else {
+        oldDestinationName = oldURL.lastPathComponent.stringByDeletingPathExtension;
+    }
+    
+    NSString *oldDestinationNameWithPathExtension = [oldDestinationName stringByAppendingPathExtension:oldURL.pathExtension];
+    
+    // If the installation and old URL differ, the sooner we can install our new app
+    // because we won't have to move the old app out of the way first
+    // increasing our chances of having it installed in case something goes wrong
+    if (![installationURL isEqual:oldURL]) {
```

Comment:
The code path that assumes they can be equal can be used for both cases. So I just did that. Like you said, when they differ is not a common case.